pkg/ottl/ottlfuncs/func_scale.go

```diff
@@ -0,0 +1,132 @@
+package ottlfuncs
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+)
+
+type ScaleArguments[K any] struct {
+	Value      ottl.Getter[K]
+	Multiplier float64
+}
+
+func NewScaleFactory[K any]() ottl.Factory[K] {
+	return ottl.NewFactory("Scale", &ScaleArguments[K]{}, createScaleFunction[K])
+}
+
+func createScaleFunction[K any](_ ottl.FunctionContext, oArgs ottl.Arguments) (ottl.ExprFunc[K], error) {
+	args, ok := oArgs.(*ScaleArguments[K])
+
+	if !ok {
+		return nil, fmt.Errorf("ScaleFactory args must be of type *ScaleArguments[K]")
+	}
+
+	return Scale(args.Value, args.Multiplier)
+}
+
+func Scale[K any](value ottl.Getter[K], multiplier float64) (ottl.ExprFunc[K], error) {
+	return func(ctx context.Context, tCtx K) (any, error) {
+		get, err := value.Get(ctx, tCtx)
+		if err != nil {
+			return nil, err
+		}
+
+		switch get.(type) {
+		case float64:
+			return get.(float64) * multiplier, nil
+		case int64:
+			return float64(get.(int64)) * multiplier, nil
+		case pmetric.NumberDataPointSlice:
+			scaledMetric := pmetric.NewNumberDataPointSlice()
+			get.(pmetric.NumberDataPointSlice).CopyTo(scaledMetric)
+			scaleMetric(scaledMetric, multiplier)
+			return scaledMetric, nil
+		case pmetric.HistogramDataPointSlice:
+			scaledMetric := pmetric.NewHistogramDataPointSlice()
+			get.(pmetric.HistogramDataPointSlice).CopyTo(scaledMetric)
+			scaleHistogram(scaledMetric, multiplier)
+			return scaledMetric, nil
+		case pmetric.ExponentialHistogramDataPointSlice:
+			scaledMetric := pmetric.NewExponentialHistogramDataPointSlice()
+			get.(pmetric.ExponentialHistogramDataPointSlice).CopyTo(scaledMetric)
+			scaleExponentialHistogram(scaledMetric, multiplier)
+			return scaledMetric, nil
+		default:
+			return nil, errors.New("unsupported data type")
+		}
+	}, nil
+}
+
+func scaleHistogram(datapoints pmetric.HistogramDataPointSlice, multiplier float64) {
+
+	for i := 0; i < datapoints.Len(); i++ {
+		dp := datapoints.At(i)
+
+		if dp.HasSum() {
+			dp.SetSum(dp.Sum() * multiplier)
+		}
+		if dp.HasMin() {
+			dp.SetMin(dp.Min() * multiplier)
+		}
+		if dp.HasMax() {
+			dp.SetMax(dp.Max() * multiplier)
+		}
+
+		for bounds, bi := dp.ExplicitBounds(), 0; bi < bounds.Len(); bi++ {
+			bounds.SetAt(bi, bounds.At(bi)*multiplier)
+		}
+
+		for exemplars, ei := dp.Exemplars(), 0; ei < exemplars.Len(); ei++ {
+			exemplar := exemplars.At(ei)
+			switch exemplar.ValueType() {
+			case pmetric.ExemplarValueTypeInt:
+				exemplar.SetIntValue(int64(float64(exemplar.IntValue()) * multiplier))
+			case pmetric.ExemplarValueTypeDouble:
+				exemplar.SetDoubleValue(exemplar.DoubleValue() * multiplier)
+			}
+		}
+	}
+}
+
+func scaleExponentialHistogram(datapoints pmetric.ExponentialHistogramDataPointSlice, multiplier float64) {
```

Comment:
I think we will also need to modify the scale of the data points. I'm not knowledgeable enough about exponential histograms to offer any insight as to how this value should be properly modified.

More on the scale here: https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto#L515

Reply:
Looking deeper into exponential histograms and how the scaling works in particular, I'm starting to doubt whether the scaling can be applied as originally intended since we cannot apply a linear factor to an exponential X axis. I guess this is also the reason exponential histograms are not supported in the [metrictransformer equivalent](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/15691) of this function.
My suggestion would be to exclude exponential histograms from this PR for now, and do this as a separate PR if we find a way of supporting this function for them. Maybe @dyladan can help here, as I've seen your article on exponential histograms [here](https://dyladan.me/histograms/2023/05/04/exponential-histograms/)?

Reply:
Thanks for the links. Thinking about it more, I agree; let's leave exponential histograms out of this.